// @nativescript/angular/esm2020/lib/legacy/router/ns-location-utils.mjs
export class Outlet {
    get isPageNavigationBack() {
        return this._navigatingBackOutlets.size > 0;
    }
    set isPageNavigationBack(isBack) {
        if (!isBack) {
            if (this._navigatingBackOutlets.size > 0) {
                this._navigatingBackOutlets.delete(this._navigatingBackOutlets.values().next().value);
            }
        }
        else {
            for (const key of this.outletKeys) {
                this._navigatingBackOutlets.add(key);
            }
        }
    }
    constructor(outletKey, path, pathByOutlets, modalNavigationDepth) {
        this._navigatingBackOutlets = new Set();
        // More than one frame available when using NSEmptyOutletComponent component
        // in module that lazy loads children (loadChildren) and has outlet name.
        this.frames = [];
        this.states = [];
        // Used in reuse-strategy by its children to determine if they should be detached too.
        this.shouldDetach = true;
        this.outletKeys = [outletKey];
        this.isPageNavigationBack = false;
        this.showingModal = false;
        this.modalNavigationDepth = modalNavigationDepth || 0;
        this.pathByOutlets = pathByOutlets;
        this.path = path;
    }
    setOutletKeyNavigatingBack(key) {
        const nests = key.split('/');
        this.outletKeys
            .filter((k) => k.split('/').length >= nests.length)
            .forEach((k) => {
            this._navigatingBackOutlets.add(k);
        });
    }
    containsFrame(frame) {
        return this.frames.indexOf(frame) > -1;
    }
    peekState() {
        if (this.states.length > 0) {
            return this.states[this.states.length - 1];
        }
        return null;
    }
    containsTopState(stateUrl) {
        const lastState = this.peekState();
        return lastState && lastState.segmentGroup.toString() === stateUrl;
    }
    // Search for frame that can go back.
    // Nested 'primary' outlets could result in Outlet with multiple navigatable frames.
    getFrameToBack() {
        let frame = this.frames[this.frames.length - 1];
        if (!this.isNSEmptyOutlet) {
            for (let index = this.frames.length - 1; index >= 0; index--) {
                const currentFrame = this.frames[index];
                if (currentFrame.canGoBack()) {
                    frame = currentFrame;
                    break;
                }
            }
        }
        return frame;
    }
}
export const defaultNavOptions = {
    clearHistory: false,
    animated: true,
};
